subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -0,0 +1,346 @@
+
+        buildFile << fixture.taskWithNestedProperty()
+        buildFile << """
+            task configureTask {
+                doLast {
+                    taskWithNestedProperty.bean = secondBean
+                }
+            }
+            
+            taskWithNestedProperty.dependsOn(configureTask)
+        """
+
+        fixture.prepareInputFiles()
+
+        when:
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        when:
+        fixture.changeFirstBean(change)
+        fixture.runTask()
+
+        then:
+        skipped(fixture.task)
+
+        when:
+        fixture.changeSecondBean(change)
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        where:
+        change << ['inputProperty', 'inputFile', 'outputFile']
+    }
+
+    @Unroll
+    def "re-configuring #change in nested bean after the task started executing has no effect"() {
+        def fixture = new NestedBeanTestFixture()
+        fixture.prepareInputFiles()
+        buildFile << fixture.taskWithNestedProperty()
+        buildFile << """
+            taskWithNestedProperty.doLast {
+                bean = secondBean
+            }
+        """
+
+        when:
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        when:
+        fixture.changeFirstBean(change)
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        when:
+        fixture.changeSecondBean(change)
+        fixture.runTask()
+
+        then:
+        skipped(fixture.task)
+
+        where:
+        change << ['inputProperty', 'inputFile', 'outputFile']
+    }
+
+    class NestedBeanTestFixture {
+        def firstInputFile = 'firstInput.txt'
+        def firstOutputFile = 'build/firstOutput.txt'
+        def secondInputFile = 'secondInput.txt'
+        def secondOutputFile = 'build/secondOutput.txt'
+
+        def task = ':taskWithNestedProperty'
+
+        def inputProperties = [
+            first: 'first',
+            second: 'second'
+        ]
+        def inputFiles = [
+            first: file(firstInputFile),
+            second: file(secondInputFile)
+        ]
+        def outputFiles = [
+            first: file(firstOutputFile),
+            second: file(secondOutputFile)
+        ]
+
+        def changes = [
+            inputProperty: { String property ->
+                inputProperties[property] = inputProperties[property] + ' changed'
+            },
+            inputFile: { String property ->
+                inputFiles[property] << ' changed'
+            },
+            outputFile: { String property ->
+                outputFiles[property] << ' changed'
+            }
+        ]
+
+        def changeFirstBean(String change) {
+            changes[change]('first')
+        }
+
+        def changeSecondBean(String change) {
+            changes[change]('second')
+        }
```

Comment:
I think we should also check for changing the bean from and to `null`, wdyt?

Reply:
I added some tests for checking this.